[bitnami/mysql] Optimize MYSQL_REPLICATION_SLAVE_DUMP behavior

mostafa-rz opened the pull request:
### Description of the change

- Remove obtain and release the database lock
- Remove finding binlog file name and current position
- Add new options to the mysqldump
  - `--single-transaction` to dump consistent data
  - `--quick` to deal with large tables
  - `--source-data=2` to add binlog file name and current position as a **_comment_** to the dump file
- Add searching the dump file to find the binlog file name and current position

### Benefits

The source database snapshot is consistent.


### Applicable issues

<!-- Enter any applicable Issues here (You can reference an issue using #) -->
- fixes #76149

mostafa-rz committed:
[bitnami/mysql] use msysqldump to lock the database, remove manual locking the database

Signed-off-by: Mostafa Rouzbahani <mostafa.rouzbahani@gmail.com>

fmulero commented:
Thanks a lot @mostafa-rz for your contribution!

I am reviewing your changes, at the moment everything looks good to me, but I'd like to have a clear scenario to test the fix. Could you provide a docker-compose and/or concrete steps to reproduce the original issue?

Thanks in advance

mostafa-rz commented:
> Thanks a lot @mostafa-rz for your contribution!
> 
> I am reviewing your changes, at the moment everything looks good to me, but I'd like to have a clear scenario to test the fix. Could you provide a docker-compose and/or concrete steps to reproduce the original issue?
> 
> Thanks in advance

Ok, let me try with docker compose file.

mostafa-rz commented:
@fmulero 
To reproduce the original issue, we need a large database to make the dump process sufficiently long. For simulation purposes, I introduced a delay in the dump process by adding a sleep command before the dump begins. This modification can be seen in the following [source code](https://github.com/bitnami/containers/blob/d8b5ecf8f55e4428a37c5dc98e28a6f672073eed/bitnami/mysql/8.4/debian-12/rootfs/opt/bitnami/scripts/libmysql.sh#L196):

```
read -r log_file log_position <<< "$(echo "SHOW BINARY LOG STATUS;" | mysql_remote_execute_print_output "$DB_MASTER_HOST" "$DB_MASTER_PORT_NUMBER" "mysql" "$DB_MASTER_ROOT_USER" "$DB_MASTER_ROOT_PASSWORD" | awk 'NR==1 {print $1, $2}')"
debug "File: $log_file and Position: $log_position"

info "Start simulate long dump process for 1 min"
sleep 60
info "End long dump simulation"

debug "Start dump process databases"

```
To test this setup, I created a custom Docker image that includes the above modification. You can access it from my [public Docker repository](https://hub.docker.com/layers/mostafabetsnap/mysql/custom-bitnami/images/sha256-210283985721fa2f6a842aac5e9b95b80eacc9f1fdf445cb56bc5a9a8fc8c704).

I used the [docker-compose-replication.yml](https://github.com/bitnami/containers/blob/main/bitnami/mysql/docker-compose-replication.yml) file to set up the containers. However, I replaced the image with my custom image from the repository.

```
# Copyright Broadcom, Inc. All Rights Reserved.
# SPDX-License-Identifier: APACHE-2.0

services:
  mysql-master:
    image: mostafabetsnap/mysql:custom-bitnami
    ports:
      - '3306'
    volumes:
      - 'mysql_master_data:/bitnami/mysql/data'
    environment:
      - MYSQL_REPLICATION_MODE=master
      - MYSQL_REPLICATION_USER=repl_user
      - MYSQL_USER=my_user
      - MYSQL_DATABASE=my_database
      # ALLOW_EMPTY_PASSWORD is recommended only for development.
      - ALLOW_EMPTY_PASSWORD=yes
      - MYSQL_ROOT_PASSWORD=my_root_password
    healthcheck:
      test: ['CMD', '/opt/bitnami/scripts/mysql/healthcheck.sh']
      interval: 15s
      timeout: 5s
      retries: 6

  mysql-slave:
    image: mostafabetsnap/mysql:custom-bitnami
    ports:
      - '3306'
    depends_on:
      - mysql-master
    environment:
      - MYSQL_REPLICATION_MODE=slave
      - MYSQL_REPLICATION_USER=repl_user
      - MYSQL_USER=my_user
      - MYSQL_DATABASE=my_database
      - MYSQL_MASTER_HOST=mysql-master
      - MYSQL_MASTER_PORT_NUMBER=3306
      - MYSQL_MASTER_ROOT_PASSWORD=my_root_password
      # ALLOW_EMPTY_PASSWORD is recommended only for development.
      - ALLOW_EMPTY_PASSWORD=yes
      # In case of missing binary files on master, use `true` to reset those binary files. Creating a previous backup is recommended.
      - MYSQL_REPLICATION_SLAVE_DUMP=true
    healthcheck:
      test: ['CMD', '/opt/bitnami/scripts/mysql/healthcheck.sh']
      interval: 15s
      timeout: 5s
      retries: 6

volumes:
  mysql_master_data:
    driver: local

```

### Steps to Reproduce the Issue

1. Run the Docker Compose file to start the MySQL master and slave containers: `docker compose -f docker-compose-replication.yml up -d`
2. Wait approximately 20 seconds for the master instance to become ready.
3. While the replica is dumping data from the master, insert some test data into the master database by running the following command(the 1 min sleep is useful here to have enough time):
```
docker exec -i mysql-mysql-master-1 mysql -u root -pmy_root_password -e "
CREATE DATABASE testDb;
USE testDb;
CREATE TABLE testTable (id INT PRIMARY KEY, name VARCHAR(255));
INSERT INTO testTable (id, name) VALUES (1, 'test');"
```

After the dump process completes, the replica fails to start. Upon inspecting the replica logs, an error is displayed that Database 'testDb' already exists. This error occurs because the replica attempts to recreate the testDb database, which was already created and populated during the initial data insertion process.

mostafa-rz commented:
Please remove the stale label.

mostafa-rz commented:
@fmulero Hi, Can you please also schedule this issue in your reviews. Many thanks.

fmulero commented on bitnami/mysql/8.4/debian-12/rootfs/opt/bitnami/scripts/libmysql.sh:
```diff
@@ -187,19 +187,21 @@ EOF
-    debug "Unlock master databases for write operations"
-    echo "UNLOCK TABLES;" | mysql_remote_execute "$DB_MASTER_HOST" "$DB_MASTER_PORT_NUMBER" "mysql" "$DB_MASTER_ROOT_USER" "$DB_MASTER_ROOT_PASSWORD"
+    # Look for the line containing "CHANGE REPLICATION SOURCE"
+    while IFS= read -r line; do
+      if [[ "$line" =~ CHANGE\ REPLICATION\ SOURCE\ TO\ SOURCE_LOG_FILE=\'([^\']+)\'\,\ SOURCE_LOG_POS=([0-9]+)\; ]]; then
+        log_file="${BASH_REMATCH[1]}"
+        log_position="${BASH_REMATCH[2]}"
+        break
+      fi
+    done < "$dump_file"
+    debug "File: $log_file and Position: $log_position"
```
If `BITNAMI_DEBUG == false` this step could consume a lot of resources for nothing. I'd check if `BITNAMI_DEBUG` is enabled before reading the dump file.

fmulero replied:
Arg! you are completely right. I've just also figure out that `SHOW MASTER STATUS` has been [deprecated](https://dev.mysql.com/doc/refman/8.4/en/show-master-status.html) in mysql 8.4 , so this feature was not running anymore. Your contribution will fix it.

Thanks

fmulero reviewed (changes requested):
Hi @mostafa-rz sorry about my very late revision.

Thank you so much for taking care of this issue and create this detailed PR. I've just left a little comment, nothing important, just to check `BITNAMI_DEBUG` before reading the dump file.

fmulero reviewed (approved):
LGTM
